### ISSUE TITLE
Cannot set AuthorizationType in CF Method (was working in RC2)

### ISSUE THREAD

**Issue by riccardo-89** [serverless/serverless#2359]
# BUG
## Description

For bug reports:
- What went wrong?
  I have updated serverless to version 1.0.2 (was working in rc2) since the update if I try to
  set the AuthorizationType to CUSTOM for using a user pool or AWS_IAM this is ignored thing that
  before was going smooth (on the other end if I set ApiKeyRequired) this goes smoothly as before.
- What did you expect should have happened?
  Setting resources props directly in cloud formation was working like charm before v1, now I think is broken in some cases.
- What was the config you used?
- What stacktrace or error message from your provider did you see?
  I do not see any errors, simply the prop is not being set.
## Additional Data
- **_Serverless Framework Version you're using**_: v1.0.2
- **_Operating System**_: mac os
- **_Stack Trace**_:
- **_Provider Error messages**_:

**Comment by flomotlik** [comment]
Can confirm, here is a config for it: 

```
functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: users/me
          method: post
      - http:
          path: users/me
          method: put

resources:
  Resources:
    ServerlessDeploymentBucket:
      Properties:
        BucketName: Testname
    ApiGatewayMethodUsersMePost:
      Properties:
        AuthorizationType: "AWS_IAM"
        Something: Something
        HttpMethod: "Whatever"
```

**Comment by eahefnawy** [comment]
@riccardo-89 thanks a lot for reporting, I looked into it and can easily reproduce. It used to work because we merged custom resources **after** compiling functions/events, but due to a recent accidental change, it's now merging custom resources **before** compiling functions/events, hence no overwrite happens.

Working on a fix now 😊

**Comment by riccardo-89** [comment]
thanks a lot guys! :D

**Comment by pmuens** [comment]
Should be resolved with #2385 /cc @flomotlik

**Comment by endymion00** [comment]
Hi,
The bug should be reopen.
When a path parameter is used it does not work.
Here is my custom vars & method definition:

```
custom:
    versionPath: v1
    usersPath: users
    cors: true

functions:
  test:
    handler: functions/users_handler.test 
    events:
        - http:
            path: ${self:custom.versionPath}/${self:custom.usersPath}/{user_id}/test
            method: post
            cors: ${self:custom.cors}
            integration: lambda

resources:
    Resources:
      ApiGatewayMethodV1UsersUseridTestPost: 
        Type: "AWS::ApiGateway::Method"
        Properties:
          HttpMethod: "POST"
          AuthorizationType: "AWS_IAM"

```

sls deploy produces this error:

Serverless: Packaging service…
Serverless: Uploading CloudFormation file to S3…
Serverless: Uploading service .zip file to S3…
Serverless: Updating Stack…
Serverless: Checking Stack update progress…
CloudFormation - UPDATE_IN_PROGRESS - AWS::CloudFormation::Stack - serverless-user-service-dev
CloudFormation - CREATE_IN_PROGRESS - AWS::ApiGateway::Method - ApiGatewayMethodV1UsersUseridTestPost
CloudFormation - **CREATE_FAILED** - AWS::ApiGateway::Method - ApiGatewayMethodV1UsersUseridTestPost
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - IamRoleLambdaExecution
CloudFormation - CREATE_IN_PROGRESS - AWS::ApiGateway::RestApi - ApiGatewayRestApi
CloudFormation - **CREATE_FAILED** - AWS::IAM::Role - IamRoleLambdaExecution
CloudFormation - **CREATE_FAILED** - AWS::ApiGateway::RestApi - ApiGatewayRestApi
CloudFormation - UPDATE_ROLLBACK_IN_PROGRESS - AWS::CloudFormation::Stack - serverless-user-service-dev
CloudFormation - UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS - AWS::CloudFormation::Stack - serverless-user-service-dev
CloudFormation - DELETE_COMPLETE - AWS::ApiGateway::RestApi - ApiGatewayRestApi
CloudFormation - DELETE_IN_PROGRESS - AWS::IAM::Role - IamRoleLambdaExecution
CloudFormation - DELETE_COMPLETE - AWS::ApiGateway::Method - ApiGatewayMethodV1UsersUseridTestPost
CloudFormation - DELETE_COMPLETE - AWS::IAM::Role - IamRoleLambdaExecution
CloudFormation - UPDATE_ROLLBACK_COMPLETE - AWS::CloudFormation::Stack - serverless-user-service-dev
Serverless: Deployment failed!

  Serverless Error ---------------------------------------

     **An error occurred while provisioning your stack: ApiGatewayMethodV1UsersUseridTestPost
     - undefined.**

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues

  Your Environment Infomation -----------------------------
     OS:                 darwin
     Node Version:       4.3.0
     Serverless Version: 1.0.3

**Comment by goldenbearkin** [comment]
I got the same error message as @endymion00 mentioned above. Please kindly reopen this thread.

**Comment by pmuens** [comment]
@endymion00 @goldenbearkin could you checkout the `master` and try again?

We've pushed a lot of fixes recently which should resolve this problem.

**Comment by mfrankwork** [comment]
@endymion00 The resource name for the API Gateway method needs to include the normalized path variable, which is converted to `<variable>Var` format by the [AWS naming library](https://github.com/serverless/serverless/blob/v1.5.0/lib/plugins/aws/lib/naming.test.js#L54-L58).  I believe your `serverless.yml` should look like the following (_HttpMethod_ is not required since it is included in the resource name):

```yaml
custom:
    versionPath: v1
    usersPath: users
    cors: true

functions:
  test:
    handler: functions/users_handler.test 
    events:
        - http:
            path: ${self:custom.versionPath}/${self:custom.usersPath}/{user_id}/test
            method: post
            cors: ${self:custom.cors}
            integration: lambda

resources:
    Resources:
      ApiGatewayMethodV1UsersUseridVarTestPost: 
        Type: "AWS::ApiGateway::Method"
        Properties:
          AuthorizationType: "AWS_IAM"
```

**Comment by saltire** [comment]
One thing I've noticed: I can change the authorization type using this method, *until* I edit it manually from the AWS API Gateway console. After that, deploying with serverless (1.5.0) has no effect on the auth type unless I remove and redeploy the stack. Something to keep in mind.

**Comment by ryanchamp-ICE** [comment]
I have a similar issue as @endymion00, where I'm using a function with a path parameter, and I try to set the authorization type as described above and I get the "HttpMethod cannot be empty" error... I am using serverless 1.6.1

**Comment by pmuens** [comment]
Thanks for posting the feedback on this one @ryanchamp-ICE 

Reopening so that we can investigate further...

**Comment by erikerikson** [comment]
We'll be coming to need this again shortly.

It'd be nice to be able to set the auth type used to access api-g either once or per "event"

**Comment by pmuens** [comment]
Thanks for the feedback @erikerikson 

Adding this to the upcoming milestone...

**Comment by ryanchamp-ICE** [comment]
I got past my issue, and I thought I'd share. One of the devs here suggested that I do a --noDeploy and find the resource name I was looking for and then set the authorizationType in the resources section as described. It worked like a charm. The key issue (at least for me) was figuring out what the name of the resource would be so I could set the auth type in resources. Apparently when you add a path variable, it attaches "Var" to the end of the normalized path, and I would have never ever guessed that lol. Looking back through this issue on the serverless forums, I think someone else did something similar to figure the name out and I just overlooked it.

Of course it would be nice if we could just set the auth type in the serverless yaml syntax where we are defining the apigateway event, but I hope this helps anyone else who runs into this problem.

**Comment by outmarch** [comment]
I used below syntax in serverless.yml and it worked for me. Key thing was to identify correct name "ApiGatewayMethodHelloGet". I figured that out from cloud formation template from S3, sls deploys code in S3 for every service if you drill down thos S3 bucket you will see cloud formation template created by sls.

resources:
Resources:
ApiGatewayMethodHelloGet:
Type: "AWS::ApiGateway::Method"
Properties:
AuthorizationType: AWS_IAM

**Comment by pmuens** [comment]
@outmarch thanks for sharing!

You can also see the generated CloudFormation templates in the `.serverless` directory which is in your service.

Furthermore there's the resources reference table in the docs which shows you how we generate the logical ids: https://serverless.com/framework/docs/providers/aws/guide/resources#aws-cloudformation-resource-reference

**Comment by HyperBrain** [comment]
@outmarch You even do not need the `Type` declaration, just the `Properties.AuthorizationType` as it gets merged into the compiled CF template during deploy.

@pmuens There is one thing missing in the resource name docs: If you have path variables in your urol, they get normalized too, *but there is a Var added* implicitly. This is a pitfall and should be documented as well there.

**Comment by pmuens** [comment]
@HyperBrain thanks for the response. I created a new issue to track that: #3343

**Comment by outmarch** [comment]
Thanks @pmuens  , @HyperBrain  good to know that same cloud formation scripts are at .serverless :)

**Comment by pmuens** [comment]
The `AWS_IAM` authorization type was added recently and will be available in v1.14. Furthermore it seems like this issue was resolved.
